torchtune/models/clip/_transforms.py

```diff
@@ -0,0 +1,176 @@
+        max_num_tiles: Optional[int] = 4,
+        resample: str = "bilinear",
+        resize_to_max_canvas: bool = False,
+    ) -> None:
+
+        # get_canvas_best_fit
+        assert (
+            possible_resolutions is not None or max_num_tiles is not None
+        ), f"Either possible_resolutions or max_num_tiles must be given. Got {possible_resolutions=} and {max_num_tiles=}"
+
+        # If possible_resolutions are not given, then calculate possible ones based on max_num_tiles
+        if not possible_resolutions and max_num_tiles:
+            possible_resolutions = find_supported_resolutions(
+                max_num_tiles=max_num_tiles, tile_size=tile_size
+            )
+        else:
+            possible_resolutions = possible_resolutions
+
+        self.possible_resolutions = torch.tensor(possible_resolutions).reshape(-1, 2)
+        logger.info(f"possible_resolutions: {self.possible_resolutions}")
+
+        self.resize_to_max_canvas = resize_to_max_canvas
+
+        # normalize
+        assert (image_mean is None) == (
+            image_std is None
+        ), f"Need to provide both or none of image_mean and image_std. Got {image_mean=} and {image_std=}"
+        self.image_mean = image_mean
+        self.image_std = image_std
+
+        # resize_with_pad
+        self.max_upscaling_size = None if resize_to_max_canvas else tile_size
+        self.resample = torchvision.transforms.InterpolationMode[resample.upper()]
+
+        # tile_crop
+        self.tile_size = tile_size
+
+    def __call__(
+        self,
+        image: Image.Image,
+    ) -> Dict[str, torch.Tensor]:
+
+        assert isinstance(image, Image.Image), "Input image must be a PIL image."
+
+        # Make image torch.tensor((3, H, W), dtype='float32'), 0<=values<=1
+        image_tensor = F.to_dtype(
+            F.grayscale_to_rgb_image(F.to_image(image)), scale=True
+        )
+
+        # Find the best canvas to fit the image without distortion
+        best_resolution = get_canvas_best_fit(
+            image=image_tensor,
+            possible_resolutions=self.possible_resolutions,
+            resize_to_max_canvas=self.resize_to_max_canvas,
+        )
+
+        # resize without distortion + pad to fit best_resolution
+        image_tensor = resize_with_pad(
+            image=image_tensor,
+            target_size=best_resolution,
+            resample=self.resample,
+            max_upscaling_size=self.max_upscaling_size,
+        )
+
+        # Normalize
+        if self.image_mean and self.image_std:
+            image_tensor = F.normalize(
+                image_tensor, mean=self.image_mean, std=self.image_std
+            )
+
+        # Divide the image into equally sized tiles
+        image_tensor = tile_crop(image=image_tensor, tile_size=self.tile_size)
+
+        aspect_ratio = torch.tensor(best_resolution).reshape(-1) // self.tile_size
+        return {
+            "image": image_tensor,
+            "ar": aspect_ratio,
```

Comment:
what will aspect ratio be used for down the pipeline?

Reply:
This is used for positional embeddings in the vision encoder: https://github.com/felipemello1/torchtune/blob/9070944e1199f047bac86e3325695bd564d35000/torchtune/models/clip/_position_embeddings.py#L153